File: Greedy/Leetcode/Medium/ValidParenthesis/ValidParenthesis.cpp
// https://leetcode.com/problems/valid-parenthesis-string/
#include<bits/stdc++.h>
using namespace std;
bool checkValidString(string s)
{
    int mini = 0, maxi = 0;
    for (int i = 0; i < s.length(); i++)
    {
        char ch = s[i];
        if (ch == '(')
        {
            mini++;
            maxi++;
        }
        else if (ch == ')')
        {
            mini--;
            maxi--;
        }
        else
        {
            mini--;
            maxi++;
        }
        mini = mini < 0 ? 0 : mini;
        if (maxi < 0)
            return false;
    }
    return mini == 0;
}
int main(void) {
    
    return 0;
}

File: Greedy/Leetcode/Medium/ValidParenthesis/ValidParenthesis.java
// https://leetcode.com/problems/valid-parenthesis-string/
public class ValidParenthesis {
    public boolean checkValidString(String s) {
        int min = 0, max = 0;
        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
            if (ch == '(') {
                min++;
                max++;
            } else if (ch == ')') {
                min--;
                max--;
            } else {
                min--;
                max++;
            }
            min = min < 0 ? 0 : min;
            if (max < 0) {
                return false;
            }
        }
        return min == 0;
    }
}
